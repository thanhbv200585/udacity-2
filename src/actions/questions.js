import { saveQuestionAnswer, saveQuestion } from "../utils/api";
import {
  addQuestionToAuthor,
  addAnsweredQuestionToUser,
} from "../actions/users";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const ADD_QUESTION = "ADD_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

function answerQuestion({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    authedUser,
    qid,
    answer,
  };
}

export function handleAnswerQuestion(info) {
  return (dispatch) => {
    dispatch(showLoading());
    dispatch(answerQuestion(info));
    dispatch(addAnsweredQuestionToUser(info));
    dispatch(hideLoading());

    return saveQuestionAnswer(info).catch((e) => {
      console.warn("Error in saveQuestionAnswer: ", e);
      alert("There was an error. Try again");
    });
  };
}

function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function handleAddQuestion(optionOne, optionTwo) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestion({
      optionOneText: optionOne,
      optionTwoText: optionTwo,
      author: authedUser,
    }).then((question) => {
      dispatch(showLoading());
      dispatch(addQuestion(question));
      dispatch(addQuestionToAuthor(question));
      dispatch(hideLoading());
    })
    .catch((e) => {
      console.warn("Error in saveQuestion: ", e);
      alert("There was an error. Try again");
    });
  };
}
