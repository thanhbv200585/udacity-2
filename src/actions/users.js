export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_ANSWERED_QUESTION = "ADD_ANSWERED_QUESTION";
export const ADD_QUESTION_TO_AUTHOR = "ADD_QUESTION_TO_AUTHOR";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function addAnsweredQuestionToUser({ qid, authedUser, answer }) {
  return {
    type: ADD_ANSWERED_QUESTION,
    qid,
    authedUser,
    answer,
  };
}

export function addQuestionToAuthor(question) {
  return {
    type: ADD_QUESTION_TO_AUTHOR,
    question,
  };
}
