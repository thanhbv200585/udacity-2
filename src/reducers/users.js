import {
  RECEIVE_USERS,
  ADD_ANSWERED_QUESTION,
  ADD_QUESTION_TO_AUTHOR,
} from "../actions/users";

export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case ADD_ANSWERED_QUESTION:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.qid]: action.answer,
          },
        },
      };
    case ADD_QUESTION_TO_AUTHOR:
      return {
        ...state,
        [action.question.author]: {
          ...state[action.question.author],
          questions: state[action.question.author].questions.concat([
            action.question.id,
          ]),
        },
      };
    default:
      return state;
  }
}
