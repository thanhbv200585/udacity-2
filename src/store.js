import reducer from "./reducers";
import { createStore } from "redux";
import middleware from "./middleware";

const store = createStore(reducer, middleware);

export { store };
