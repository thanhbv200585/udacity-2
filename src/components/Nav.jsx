import { Link, NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import clsx from "clsx";

const NAVS = [
  {
    path: "/",
    label: "Home",
  },
  {
    path: "/add",
    label: "Add Poll",
  },
  {
    path: "/leaderboard",
    label: "Leader Board",
  },
];

const Nav = (props) => (
  <nav className="flex justify-between items-center gap-3 h-12">
    <ul className="flex gap-4">
      {NAVS.map((nav) => (
        <li key={nav.label}>
          <NavLink
            to={nav.path}
            className={({ isActive }) =>
              clsx(
                "transition-all relative",
                "after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all",
                "hover:text-blue-500 hover:after:bg-blue-400",
                isActive
                  ? "text-blue-500 after:bg-blue-400"
                  : "after:bg-transparent"
              )
            }
          >
            {nav.label}
          </NavLink>
        </li>
      ))}
    </ul>
    <ul className="flex items-center gap-4">
      <li>
        <Link
          onClick={(e) => props.dispatch(setAuthedUser(null))}
          className="hover:text-blue-500 hover:after:bg-blue-400"
        >
          Logout
        </Link>
      </li>
      <li className="flex items-center gap-2">
        <img
          hidden={props.name === undefined}
          alt={`avatar of ${props.name}`}
          src={props.avatarURL}
          className="w-8 h-8 rounded-full"
        />
        <span>{props.name}</span>
      </li>
    </ul>
  </nav>
);

const mapStateToProps = ({ authedUser, users }) => {
  const avatarURL = users[authedUser]?.avatarURL;
  const name = users[authedUser]?.name;

  return {
    name,
    avatarURL,
  };
};

export default connect(mapStateToProps)(Nav);
