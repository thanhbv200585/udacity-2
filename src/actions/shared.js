import { getInitialData } from "../utils/api";
import { receiveQuestions } from "./questions";
import { receiveUsers } from "./users";
import { setAuthedUser } from "./authedUser";
import { showLoading, hideLoading } from "react-redux-loading-bar";

const AUTHED_ID = null

export function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveQuestions(questions));
      dispatch(receiveUsers(users));
      dispatch(setAuthedUser(AUTHED_ID));
      dispatch(hideLoading());
    });
  };
}
