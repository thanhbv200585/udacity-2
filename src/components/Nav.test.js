import { render, screen } from "@testing-library/react";
import Nav from "./Nav";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { setAuthedUser } from "../actions/authedUser";
import { receiveUsers } from "../actions/users";

describe("Nav Component Tests", () => {
  it("should render the Nav component", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Nav />
        </Router>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should display user's name and avatar when authenticated", async () => {
    const mockUser = {
      mtsamis: {
        id: "mtsamis",
        name: "Mike Tsamis",
        avatarURL: "https://www.shutterstock.com/image-vector/young-smiling-man-adam-avatar-600w-2107967969.jpg",
      },
    };

    await store.dispatch(receiveUsers(mockUser));
    await store.dispatch(setAuthedUser("mtsamis"));
    
    render(
      <Provider store={store}>
        <Router>
          <Nav />
        </Router>
      </Provider>
    );

    expect(screen.getByText("Mike Tsamis")).toBeInTheDocument();
    expect(screen.getByAltText("Mike Tsamis's avatar")).toBeInTheDocument();
  });
});
