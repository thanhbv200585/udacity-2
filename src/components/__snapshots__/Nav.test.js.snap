// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav is expected to render the Nav component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="flex justify-between items-center gap-3 h-12"
      >
        <ul
          class="flex gap-4"
        >
          <li>
            <a
              aria-current="page"
              class="transition-all relative after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all hover:text-blue-500 hover:after:bg-blue-400 text-blue-500 after:bg-blue-400"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="transition-all relative after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all hover:text-blue-500 hover:after:bg-blue-400 after:bg-transparent"
              href="/add"
            >
              Add Poll
            </a>
          </li>
          <li>
            <a
              class="transition-all relative after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all hover:text-blue-500 hover:after:bg-blue-400 after:bg-transparent"
              href="/leaderboard"
            >
              Leader Board
            </a>
          </li>
        </ul>
        <ul
          class="flex items-center gap-4"
        >
          <li>
            <a
              class="hover:text-blue-500 hover:after:bg-blue-400"
              href="/"
            >
              Logout
            </a>
          </li>
          <li
            class="flex items-center gap-2"
          >
            <img
              alt="avatar of undefined"
              class="w-8 h-8 rounded-full"
              hidden=""
            />
            <span />
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="flex justify-between items-center gap-3 h-12"
    >
      <ul
        class="flex gap-4"
      >
        <li>
          <a
            aria-current="page"
            class="transition-all relative after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all hover:text-blue-500 hover:after:bg-blue-400 text-blue-500 after:bg-blue-400"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class="transition-all relative after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all hover:text-blue-500 hover:after:bg-blue-400 after:bg-transparent"
            href="/add"
          >
            Add Poll
          </a>
        </li>
        <li>
          <a
            class="transition-all relative after:absolute after:bottom-[-2px] after:left-0 after:w-full after:h-0.5 after:transition-all hover:text-blue-500 hover:after:bg-blue-400 after:bg-transparent"
            href="/leaderboard"
          >
            Leader Board
          </a>
        </li>
      </ul>
      <ul
        class="flex items-center gap-4"
      >
        <li>
          <a
            class="hover:text-blue-500 hover:after:bg-blue-400"
            href="/"
          >
            Logout
          </a>
        </li>
        <li
          class="flex items-center gap-2"
        >
          <img
            alt="avatar of undefined"
            class="w-8 h-8 rounded-full"
            hidden=""
          />
          <span />
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
